systemd: 254.6 -> 255.2

nikstur opened the pull request:
## Description of changes

- Enable password quality by default. We could have done this a long time ago via libpwquality. systemd only introduced  passwdqc as an alternative over libpwquality in v254, but it still prefers libpwquality.
- Enable qrencode by default. This also enables building systemd-bsod.
- Convert mesonFlags to use `lib.meson*` options to remove deprecation warnings and reorganize them.
- Enable vmspawn by default (this is a new experimental component in v255)

Upstream Changelog: https://github.com/systemd/systemd/releases/tag/v255

I discussed with @RaitoBezarius about creating a new `systemd-next` attribute that contains the release candidate of the next systemd release. Although I agree with him that this attribute would make testing easier, I think it will be a lot of work to maintain both versions. 

For this release I decided against the systemd-next route. But I would want to pursue this for the next systemd release since I have been convinced that the benefits outweigh the costs.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented:
> Although I agree with him that we need this attribute for better testing I think it will be a lot of work to maintain both versions.

Normally we do not add prerelease versions exactly because of this reason unless the last release is really far in the past. Also see the recent removal of chromium/chrome dev/beta.

RaitoBezarius commented:
My belief is that we have 4-5 systems maintainers around here, it's not the same as chromium.

systemd is also on the critical path.

Also, we have the option to not make this built by Hydra either case.

lilyinstarlight commented:
Shouldn't this target `staging`?

ElvishJerricco commented:
I haven't built this myself yet to check, but the release notes for systemd 255 say that `systemd-cryptsetup` was moved to `$out/bin`, so we'll have to update the path that's pulled from in systemd stage 1.

nikstur commented:
> Shouldn't this target `staging`?

It should.

For early testing, however, targeting master is easier because more things are cached afaik. 

If we go the systemd-next route, we can merge this into master.

nikstur committed:
systemd: remove unused bindings

nikstur committed:
systemd: 254.6 -> 255.2

Removed patches:

- 0007-Fix-hwdb-paths.patch

  The directory we want seems to already be included in the list. Is there
  a reason why we want to restrict it further?

- 0010-build-don-t-create-statedir-and-don-t-touch-prefixdi.patch

  This patch has little to do with how the meson.build file looks now. The
  new patch 0017 is the successor to this one.

- 0015-pkg-config-derive-prefix-from-prefix.patch

  This is fixed upstream. We don't need this anymore.

jtojnar commented on pkgs/development/libraries/pipewire/default.nix:
```diff
@@ -189,6 +189,7 @@ let
       "-Dsdl2=disabled" # required only to build examples, causes dependency loop
       "-Drlimits-install=false" # installs to /etc, we won't use this anyway
       "-Dcompress-offload=enabled"
+      "-Dsystemd-system-unit-dir=${placeholder "out"}/lib/systemd"
```
[This change](https://github.com/systemd/systemd/commit/b0d3095fd6cc1791a38f57a1982116b4475244ba#diff-3dce99316f44198c53d8147020aaca9c8931b5248587c64c098fdcef8ed4da03) appears to cause a regression in overridability, which we previously restored in https://github.com/systemd/systemd/commit/60bce7c6d9606185114df1bdcd5ea100407688b8.

Many projects have been overriding `root_prefix` and will now be broken:
https://gitlab.freedesktop.org/mobile-broadband/ModemManager/-/blob/e35655b364d32fce787cd3980c876f3946f82192/meson.build#L204

ElvishJerricco replied:
Yes, we were discussing this on matrix today. Ideally, we would fix every package in the world to handle this correctly, instead of in the flawed way they do it now with `root(_)prefix`. But it's simpler to just make systemd do things the weird bad way:

```diff
diff --git a/src/core/systemd.pc.in b/src/core/systemd.pc.in
index 6cca2fad9a..8e39c5247e 100644
--- a/src/core/systemd.pc.in
+++ b/src/core/systemd.pc.in
@@ -11,9 +11,9 @@
 # considered deprecated (though there is no plan to remove them). New names
 # shall have underscores.
 
-prefix={{PREFIX_NOSLASH}}
-root_prefix=${prefix}
-rootprefix=${prefix}
+root_prefix={{PREFIX_NOSLASH}}
+rootprefix=${root_prefix}
+prefix=${rootprefix}
 sysconf_dir={{SYSCONF_DIR}}
 sysconfdir=${sysconf_dir}
```

This patch seems to be fixing builds for me, though it's really hard to test because my personal hydra is really unhappy doing world-rebuilds like this and I'm finding all sorts of nondeterministic build failures that are getting in the way of testing that this actually works -_-

jtojnar replied:
Well, `--define-variable` is [standard part of pkg-config](https://www.bassi.io/articles/2018/03/15/pkg-config-and-paths/) so I would argue https://github.com/systemd/systemd/commit/b0d3095fd6cc1791a38f57a1982116b4475244ba broke a public API.

But in the end, it depends on systemd’s policy – maybe they do not consider this part of their API promise. Though, they [accepted our fix](https://github.com/systemd/systemd/pull/18118) when they broke it in the past.

Personally, I would prefer applying your patch upstream and announcing the `root_prefix` for pkg-config override is deprecated, so that we have some time to fix it in projects installing service files.

Unfortunately, meson currently [does not support multiple `define_variable`](https://github.com/mesonbuild/meson/pull/10670) so projects will need to do something like this:

https://github.com/fwupd/fwupd/blob/b678ebeb0ae57c9b9c335db262996c4a8e8ff5c4/meson.build#L465-L469

nikstur replied:
AFAIU this is fixed by https://github.com/systemd/systemd/pull/30115 in 255-rc3

clhodapp commented:
I'll drop in here that ukify is non-experimental in 255. It would be super-nice to get a packaged ukify somewhere in nixpkgs at this point (`systemd-ukify` with python in its closure, a dedicated `ukify` package, etc.).

lilyinstarlight commented:
> It would be super-nice to get a packaged ukify somewhere in nixpkgs at this point (`systemd-ukify` with python in its closure, a dedicated `ukify` package, etc.).

You can use it now actually via `systemd.override { withUkify = true; }` but it is disabled by default

I wonder if we could start building it for the non-minimal systemd build by default though and just put it in its own `ukify` output (e.g. accessible as `systemd.ukify`) so that way it doesn't end up in the runtime closure unless something uses it?

clhodapp commented:
That would wonderful. The issue with overriding `withUkify`  to `true` is that there isn't currently a reference to that from any `all-packages` attributes so it doesn't end up in the cache.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -32265,6 +32269,15 @@ with pkgs;
 
   qrencode = callPackage ../development/libraries/qrencode {
     inherit (darwin) libobjc;
+    # Needed to break dependency cycles in systemd. SDL2 is only needed for
+    # tests in qrencode. This also reduces the number of rebuilds for systemd.
+    SDL2 = SDL2.override {
+      udevSupport = false;
+      dbusSupport = false;
+      pipewireSupport = false;
+      libdecorSupport = false;
+      pulseaudioSupport = false;
+    };
```
Or simpler: we just copy config/no-x-libs and disable the check entirely and move it to passthru

nikstur replied:
Good idea!

nikstur commented:
Now targets staging. I will not pursue the systemd-next route for this release but I am interested in doing it for the next release.

nikstur commented:
I'm currently testing this on my machine. I created a branch that bases these changes on nixos-unstable so it is easier to generate a patch set that applies cleanly: https://github.com/nikstur/nixpkgs/tree/systemd-255-unstable

You can use this to test locally.

ElvishJerricco reviewed (changes requested):
@nikstur This is looking good to me. Let's merge this very soon. But first, can we bump to 255.2?

nikstur committed:
qrencode: move test into passthru

This is needed to break a dependency for systemd.

nikstur commented:
>  But first, can we bump to 255.2?

Done

Changes of the last force push:

- Use finalAttrs to refer to the qrencode package instead of pulling it in as a callPackage input
- Inlcude the timesync patch that was on staging and now has reached master

nikstur commented:
I don't understand why the ofborg `Check whether nix files are parseable` job fails. Can someone help me out here?

wahjava commented:
> I don't understand why the ofborg `Check whether nix files are parseable` job fails. Can someone help me out here?

    | jq '.[] | select(.status != "removed" and (.filename | endswith(".nix"))) | .filename' \

in the preceding step of the task outputs quoted file names, e.g. `"nixos/modules/system/boot/luksroot.nix"` which are passed to `nix-instantiate` verbatim. To reproduce:

    ❯ file='"nixos/modules/system/boot/luksroot.nix"'
    ❯ foo="$(nix-instantiate --parse "$file")"
    error: path '/nixpkgs/"nixos/modules/system/boot/luksroot.nix"' does not exist

For a fix, `s/jq/jq -r/`

nikstur commented:
Ah ok, so this has nothing to do with my changes. This commit fixes it on master: https://github.com/NixOS/nixpkgs/commit/2ccf45ee95e4ceba8365837d773e284030e28d6f

This commit probably won't reach stagin anytime soon. So it should be safe to ignore it.

SuperSandro2000 reviewed (approved):
I tested 255.0 on WSL2 without any issues.

RaitoBezarius commented:
Once CI is green, I will merge.

JohnRTitor commented:
Looks like `systemd-bsod` support was part of the v255 release, and v256 is around the corner, but I don't see a module/option to enable this feature. Am I missing something or this feature is not yet available on NixOS?

arianvp commented:
Contributions welcome. We don't enable all systemd features by default.
